storage/remote/queue_manager.go

```diff
@@ -688,41 +727,77 @@ func (s *shards) start(n int) {
 	s.qm.metrics.numShards.Set(float64(n))
 
 	newQueues := make([]chan sample, n)
+	buffer := make([]chan []sample, n)
 	for i := 0; i < n; i++ {
 		newQueues[i] = make(chan sample, s.qm.cfg.Capacity)
+		buffer[i] = make(chan []sample, s.qm.cfg.Capacity)
 	}
 
+	if s.queues != nil {
+		// newQueues consume samples in the existing queues followed by the. Once
+		// the samples from existing queues are consumed, they will be replaced by
+		// these queues which will then ingest the incoming samples.
+		s.newQueuesAvailable.Store(true)
+		s.newQueueBuf = &buffer
+	} else {
+		s.queueBuf = &buffer
+	}
 	s.queues = newQueues
 
 	var hardShutdownCtx context.Context
 	hardShutdownCtx, s.hardShutdown = context.WithCancel(context.Background())
-	s.softShutdown = make(chan struct{})
 	s.running.Store(int32(n))
+	s.reshard = make(chan struct{})
 	s.done = make(chan struct{})
 	s.droppedOnHardShutdown.Store(0)
 	for i := 0; i < n; i++ {
 		go s.runShard(hardShutdownCtx, i, newQueues[i])
 	}
 }
 
-// stop the shards; subsequent call to enqueue will return false.
+// stop the shards. This stops existing shards and triggers flushing of samples
+// that are already existing in the queue.
 func (s *shards) stop() {
-	// Attempt a clean shutdown, but only wait flushDeadline for all the shards
-	// to cleanly exit.  As we're doing RPCs, enqueue can block indefinitely.
-	// We must be able so call stop concurrently, hence we can only take the
-	// RLock here.
-	s.mtx.RLock()
-	close(s.softShutdown)
-	s.mtx.RUnlock()
+	for _, queue := range s.queues {
+		close(queue)
+	}
+	s.flushRemaining(*s.queueBuf)
+}
 
+// flushRemaining flushes the remaining samples from the buffer that were in pending state
+// in the queue.
+func (s *shards) flushRemaining(buffer []chan []sample) {
+	var (
+		buf       []byte
+		tempStore []sample
+	)
+	for i := 0; i < len(buffer); i++ {
+		buf := <-buffer[i]
+		if buf == nil {
+			continue
+		}
+		tempStore = append(tempStore, buf...)
+	}
+	if len(tempStore) == 0 {
+		return
+	}
+	sort.SliceStable(tempStore, func(i, j int) bool {
+		return tempStore[i].t < tempStore[j].t
+	})
+	level.Info(s.qm.logger).Log("msg", "Flushing samples before stopping shards", "count", len(tempStore))
+	s.sendSamples(context.Background(), createTimeSeriesFromSamples(tempStore), &buf)
+}
+
+// stop the old shards; subsequent call to enqueue will return false.
+func (s *shards) stopOldShards() {
```

Comment:
Having `stop` and `stopOldShards` is again confusing naming IMO. Can we can name this one something like `stopForResharding` or something to that effect?

Reply:
After the recent push, I feel that `stopOldShards()` is not required anymore.